modules/shortcodes/instagram.php

```diff
@@ -78,112 +78,91 @@ function jetpack_instagram_embed_reversal( $content ) {
 wp_oembed_remove_provider( '#https?://(www\.)?instagr(\.am|am\.com)/(p|tv)/.*#i' );
-wp_embed_register_handler(
-	'jetpack_instagram',
-	'#http(s?)://(www\.)?instagr(\.am|am\.com)/(p|tv)/([^\/]*)#i',
-	'jetpack_instagram_handler'
+
+wp_oembed_add_provider(
+	'#https?://(www\.)?instagr(\.am|am\.com)/(p|tv)/.*#i',
+	'https://graph.facebook.com/v5.0/instagram_oembed/',
+	true
 );
 
 /**
- * Handle Instagram embeds (build embed from regex).
- *
- * @param array  $matches Array of matches from the regex.
- * @param array  $atts    The original unmodified attributes.
- * @param string $url     The original URL that was matched by the regex.
+ * Handle an alternate Instagram URL format, where the username is also part of the URL.
  */
-function jetpack_instagram_handler( $matches, $atts, $url ) {
-	global $content_width;
-
-	$max_width = 698;
-	$min_width = 320;
+wp_oembed_add_provider(
+	'#https?://(?:www\.)?instagr(?:\.am|am\.com)/(?:[^/]*)/(p|tv)/([^\/]*)#i',
+	'https://graph.facebook.com/v5.0/instagram_oembed/',
+	true
+);
 
-	if ( is_feed() ) {
-		// Instagram offers direct links to images, but not to videos.
-		if ( 'p' === $matches[1] ) {
-			$media_url = sprintf( 'https://instagr.am/p/%1$s/media/?size=l', $matches[2] );
-			return sprintf(
-				'<a href="%1$s" title="%2$s" target="_blank"><img src="%3$s" alt="%4$s" /></a>',
-				esc_url( $url ),
-				esc_attr__( 'View on Instagram', 'jetpack' ),
-				esc_url( $media_url ),
-				esc_html__( 'Instagram Photo', 'jetpack' )
-			);
-		} elseif ( 'tv' === $matches[1] ) {
-			return sprintf(
-				'<a href="%1$s" title="%2$s" target="_blank">%3$s</a>',
-				esc_url( $url ),
-				esc_attr__( 'View on Instagram', 'jetpack' ),
-				esc_html__( 'Instagram Video', 'jetpack' )
-			);
-		}
+/**
+ * Add auth token required by Instagram's oEmbed REST API, or proxy through WP.com.
+ *
+ * @since 9.1.0
+ *
+ * @param string $provider URL of the oEmbed provider.
+ * @param string $url      URL of the content to be embedded.
+ *
+ * @return string
+ */
+function jetpack_instagram_oembed_fetch_url( $provider, $url ) {
+	if ( ! wp_startswith( $provider, 'https://graph.facebook.com/v5.0/instagram_oembed/' ) ) {
+		return $provider;
 	}
 
-	$atts = shortcode_atts(
-		array(
-			'width'       => isset( $content_width ) ? $content_width : $max_width,
-			'hidecaption' => false,
-		),
-		$atts
-	);
-
-	$atts['width'] = absint( $atts['width'] );
-	if ( $atts['width'] > $max_width ) {
-		$atts['width'] = $max_width;
-	} elseif ( $atts['width'] < $min_width ) {
-		$atts['width'] = $min_width;
+	// Attempt to clean query params from the URL since Facebook's new API for Instagram
+	// embeds does not like query parameters. See p7H4VZ-2DU-p2.
+	$parsed_url = wp_parse_url( $url );
+	if ( $parsed_url ) {
+		$clean_url = 'https://www.instagram.com' . $parsed_url['path'];
+		$provider  = add_query_arg( 'url', rawurlencode( $clean_url ), remove_query_arg( 'url', $provider ) );
```

Comment:
This `remove_query_arg` inside `add_query_arg` made me do a double take when I first read it :)

Reply:
I can spread it over two lines (with an intermediate `$provider_without_url` or `$provider_without_emned_url` var), and/or add a comment :slightly_smiling_face:

Reply:
d117c6571

Reply:
Thank you!